source/extensions/filters/http/header_mutation/header_mutation.cc

```diff
@@ -63,15 +61,16 @@ Http::FilterHeadersStatus HeaderMutation::encodeHeaders(Http::ResponseHeaderMap&
   config_->mutations().mutateResponseHeaders(request_headers, headers,
                                              encoder_callbacks_->streamInfo());
 
-  if (route_config_ == nullptr) {
-    // If we haven't already resolved the route config, do so now.
-    route_config_ = Http::Utility::resolveMostSpecificPerFilterConfig<PerRouteHeaderMutation>(
-        encoder_callbacks_);
+  // If we haven't already traversed the route configs, do so now.
+  if (route_configs_.empty()) {
+    route_configs_ =
+        Http::Utility::getAllPerFilterConfig<PerRouteHeaderMutation>(encoder_callbacks_);
   }
 
-  if (route_config_ != nullptr) {
-    route_config_->mutations().mutateResponseHeaders(request_headers, headers,
-                                                     encoder_callbacks_->streamInfo());
+  for (const auto* route_config : route_configs_) {
+    ASSERT(route_config != nullptr);
+    route_config->mutations().mutateResponseHeaders(request_headers, headers,
```

Comment:
You probably want to have some comment to talk about order here, e.g. are we respecting most specific to least specific ordering of the config?

Reply:
I think this deserve a config flag. Just like the `most_specific_header_mutations_wins` in the route configuration.

Reply:
Yea, i have thought about specific ordering and plan to use default orders here with TODO for reverse order (as I mentioned in PR description).

On second thought, I think just adding config in this PR make sense. Will do.